src/libsystemd/sd-event/event-util.c

```diff
@@ -0,0 +1,70 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include "event-util.h"
+#include "log.h"
+
+int event_reset_time(sd_event *e, sd_event_source **s, clockid_t clock, uint64_t usec, uint64_t accuracy, sd_event_time_handler_t callback, void *userdata, bool force_reset) {
+        int enabled, r;
+        clockid_t c;
+
+        assert(e);
+        assert(s);
+
+        if (!*s)
+                goto create_new;
+
+        if (!force_reset) {
+                r = sd_event_source_get_enabled(*s, &enabled);
+                if (r < 0) {
+                        log_debug_errno(r, "sd-event: Failed to query whether event source is enabled or not, trying to create new event source: %m");
+                        goto create_new;
+                }
+
+                if (enabled != SD_EVENT_OFF)
+                        return 0;
+        }
+
+        r = sd_event_source_get_time_clock(*s, &c);
+        if (r < 0) {
+                log_debug_errno(r, "sd-event: Failed to get clock id, trying to create new event source: %m");
+                goto create_new;
+        }
+
+        if (c != clock) {
+                log_debug("Current clock id is different from specified one, trying to create new event source.");
+                goto create_new;
+        }
+
+        r = sd_event_source_set_time(*s, usec);
+        if (r < 0) {
+                log_debug_errno(r, "sd-event: Failed to set time to event source, trying to create new event source: %m");
+                goto create_new;
+        }
+
+        r = sd_event_source_set_time_accuracy(*s, accuracy);
+        if (r < 0) {
+                log_debug_errno(r, "sd-event: Failed to set accuracy, trying to create new event source: %m");
+                goto create_new;
+        }
+
+        /* callback function is not updated, as we do not have sd_event_source_set_time_callback(). */
+
+        (void) sd_event_source_set_userdata(*s, userdata);
+
+        r = sd_event_source_set_enabled(*s, SD_EVENT_ONESHOT);
+        if (r < 0) {
+                log_debug_errno(r, "sd-event: Failed to enable event source, trying to create new event source: %m");
+                goto create_new;
+        }
+
+        return 1;
+
+create_new:
+        *s = sd_event_source_unref(*s);
+
+        r = sd_event_add_time(e, s, clock, usec, accuracy, callback, userdata);
+        if (r < 0)
+                return log_warning_errno(r, "sd-event: Failed to create timer event: %m");
+
+        return 2;
```

Comment:
not a fan of such magic return values... if mean, it's fine to return "extended booleans", i.e. booleans that are true, false or "failure". But in this case you return more than two "true" values...

I think in this case we should really add an enum (can be anonymous if necessary), to give this explanatory names.

Thaid said, is there are actually need to distuingish the two cases of "i updated an existing source" and "i left the existing source untouched"? if not, then it might make sense to return 0 in both those cases, and then return > 0 only in the case of "i allocated a new event source for you"